RedHatInsights/insights-rbac
# Add patch support to role detail

## Link(s) to Jira
- https://issues.redhat.com/browse/RHCLOUD-10938

## Description of Intent of Change(s)
>We should support an explicit PATCH method for /roles/ in order to allow for the ability to update the "name" and "description" fields ad-hoc, without having to supply the entire payload, including "access".
>
>This is a requirement that the FE is requesting, otherwise they'd need to either query /roles/uuid/ from the /roles/ page in the UI when someone wants to update the name or description, since currently the update method requires "access" be supplied, or would require we add the access data to /roles/ which would be a performance hit, and not terribly resourceful.
>
>Instead we should consider having the FE submit a PATCH request to /roles/uuid/ with the "name" and or "description" field supplied.

This change allows for a `PATCH` request to be made against `/roles/<uuid>/` without
needing to supply the entire `Role` instance payload, along with the `access` object
required in the `PUT`.

Instead, you can now supply a `PATCH` with any/all of the following:
```json
{
  "name": "Name of a role",
  "display_name": "Display name of a role",
  "description": "Description of a role"
}
```

This will then return the entire role object, the same as a `GET` or `PUT` would.

This makes use of the `partial_update` method, which skips the access validation,
validates the UUID and then updates the role using a new serializer `RolePatchSerializer`
which is only used on `PATCH` requests to `/roles/<uuid>`. This serializer inherits
from the `RoleSerializer` but overrides the `access` to be optional, and does not
validate `access` on update, explicitly setting the `name`, `display_name` and
`description` fields only.

## Local Testing
To test this locally:
- submit the following payload as a `PUT` on a role and see how validation fails
  because you currently need to send the entire payload:
  ```
    {
      "name": "Name of a role",
      "display_name": "Display name of a role",
      "description": "Description of a role"
    }
  ```
- submit the same payload as a `PATCH` and see that it succeeds and returns the role


## Checklist
- [x] if API spec changes are required, is the spec updated?

- [ ] are there any pre/post merge actions required? if so, document here.

- [x] are theses changes covered by unit tests?

- [ ] if warranted, are documentation changes accounted for?

- [ ] does this require migration changes?
  - [ ] if yes, are they backwards compatible?

- [ ] is there known, direct impact to dependent teams/components?
  -  if yes, how will this be handled?

## Secure Coding Practices Checklist Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Practices Checklist
